Expand test suite to all examples and range of sizes

GMNGeoffrey opened the pull request:
Rather than relying on funky shell script parsing, we just write out the test cases as yaml configs directly. This also means that we can optimize for being a test rather than being a good user example, for instance, by making the symmetry tests way smaller.

But testing larger lengths is still useful, especially for benchmarking, so we use the base configs and then programmatically extend them to generate larger configs. These large tests are excluded by default in pytest. Small tests now take about 90s for me serially, whereas when I started I think they were on the order of 600 (some of that is other improvements, like doing everything in-process, I believe). So these are still very much end-to-end tests, not unit tests, but much friendlier to run.

All the test configs explicitly list their length, which is validated by `ContigMap`.

Note that I took a look at the contig parsing code, and it turns out that you can just list a single number when the contig is a fixed length rather than a range, so "150" instead of "150-150": https://github.com/Brium/RFdiffusion/blob/7cd3ac12b0/rfdiffusion/contigs.py#L214-L217. I used that format throughout, except for scaffold guided insertions where it instead means 0-150, somewhat surprisingly.

I made some additional fixes and error-checking improvements. I may break those off into separate PRs:
1. Report the defined and sampled lengths when complaining that contigs and length are incompatible.
2. Do the above validation even when contigs is determined automatically in the scaffold guided case.
3. Actually make it possible to specify scaffold_list in the config.
4. Don't print a bunch of whitespace in the middle of some error messages (this was using line continuations somewhat inappropriately).
5. Make pytest not blackhole info logs, which RFdiffusion uses kind of interchangeably with print statements.

GMNGeoffrey committed:
Translate the rest of the example scripts

Fixed lengths help a lot with determinism. I gave all the configs nice
names also. I took a look at the contig parsing code, and it turns out
that you can just list a single number when the contig is a fixed length
rather than a range, so "150" instead of "150-150":

https://github.com/Brium/RFdiffusion/blob/7cd3ac12b0/rfdiffusion/contigs.py#L44-L48

GMNGeoffrey commented:
@Joachim-Sh could you take a look?

Joachim-Sh reviewed (commented):
Expanding the tests is a good idea. What do you think of adding a "mark" to the larger tests? So users can choose to run all tests or only the "smaller" ones. If yes, we could also increase the number of AA generated into the 100s like you suggested.

GMNGeoffrey commented:
Yeah, pytest has good facilities for marking tests. I'll take a look. If we do that, do you think we should extend the ranges above the max used in the examples? Is there anything special about the lengths used in the examples?

Joachim-Sh commented:
Yes, extending above the max ranges is a good idea for testing/benchmarking. In real life, you are probably not generating a huge protein around a small motif, but for testing/benchmarking, it is fine.

GMNGeoffrey committed:
Generate configs of various lengths

GMNGeoffrey commented:
Ok, I wrote a script to start from the config for short designs and extend them. I wasn't sure what to do for the partial diffusion or tim barrel cases. I figure that for unit tests we mostly just want these very short ones (even shorter than the example scripts in some cases), but for benchmarking we'll want a range of lengths. I went up to a length of 1200, since one of the designs went up that high, though I think maybe it's only workable at that length due to symmetry? WDYT @Joachim-Sh? If this general thing looks good, I'll make it so that it's easy to run either only the short tests or all of them and then set up a comprehensive benchmarking suite for the full thing.

Joachim-Sh commented:
Yes indeed, I would only go up to 1200 for the symmetry cases, for the other ones I would stick with 700. What would be the expected runtime for the full benchmark?

GMNGeoffrey commented:
I was thinking of going up to 1000 still, since that's where the graphs in the original paper stop and it's a nice round number, but we can do 700 instead if you think that's better (the paper does say that unconditional designs are only really good up to 600aa).

For full benchmark runtime, it depends on what you're asking:
1. using the per-step benchmarking to measure per-step latency for each design
2. running every full design once for an end-to-end time
3. running each full design enough times to have stable results individually

I'd guess about 4, 8, and 48 GPU hours, respectively (this is still using 1k as an upper bound for the non-symmetry cases and not adding any more partial diffusion or tim barrel test cases).

Do you have ideas on how to expand the partial diffusion and tim barrel cases?

GMNGeoffrey committed:
Separate small and large generated configs

Large tests are all longer-chain versions of the small tests. By
default, they aren't run because we set `-m "not large"` in pytest.ini.
To run them you can pass `-m ""` or to run only them `-m "large"`.

At this point, the tests that directly look at the example bash scripts
aren't really doing anything, so I removed them.

The generation script isn't my cleanest, but it does work.

GMNGeoffrey committed:
Fix logging setting

I was trying to figure out how to make pytest output logs interleaved
with stdout, as that would match the normal output when running
intference since RFdiffusion freely mixes logging and prints. I couldn't
get that to work though, so the best we can do is actually show the info
logs by default. If more of the RFdiffusion output were put into logs
instead of prints then it would be more shared again.

GMNGeoffrey committed:
Avoid symlinking example files into the test dir

We can just run the tests with the correct working directory and we now
send their output to a temporary directory via an absolute path, so
there's no concern about writing into the example directory.

GMNGeoffrey committed:
Improve error message for contig/length mismatch

This was helpful when setting up example test cases. This could actually
be validated statically rather than doing 100k attempts, but that's a
separate issue and would maybe be more cumbersome than it's worth.

Joachim-Sh commented:
@GMNGeoffrey Oke nice, that is fewer GPU hours than I expected. 

For the partial diffusion, I would fully diffuse some of the smaller and larger input_pbs: 

- tau_peptide.pdb: 178 residues
- **5TPN.pdb: 916 residues**
- 3IOL.pdb: 126 residues
- 2KL8.pdb: 79 residues
- 1YCR.pdb: 98 residues
- **5an7.pdb: 242 residues**
- 1qys.pdb: 91 residues
- nickel_symmetric_motif.pdb: 20 residues
- peptide_complex_ideal_helix.pdb: 206 residues
- insulin_target.pdb: 150 residues

For the tim_barrel example, it would make most sense to increase the number of residues to the N and the C-terminus: scaffoldguided.sampled_N=0-5 and scaffoldguided.sampled_C=0-5. They do not need to be the same and can be increased as much as we want.

GMNGeoffrey commented:
SG for partial diffusion. I'll add 5an7.pdb and 5TPN.pdb. I've already got a tim barrel example `timbarrel_long` that sets all those parameters to 5, but the resulting pdb still only has 227 residues. Maybe we just don't have any more examples for that 🤷

Joachim-Sh commented:
For the tim_barrel example, you could increase the number of residues of the N and the C-terminus to e.g. 100 or 500. These are the parameters scaffoldguided.sampled_N and scaffoldguided.sampled_C. They do not need to be the same and can be increased as much as we want.

GMNGeoffrey commented:
Oh got it, thanks for clarifying. Reading comprehension fail on my part 😁

GMNGeoffrey commented:
What about `scaffoldguided.sampled_insertion`. Looks like this one is added to every loop, so maybe good reason to keep it small?

Joachim-Sh commented:
Yes indeed. Loops are typically quite small, so I would leave that parameter as is. As opposed to generating larger protein at the end (N and C terminus), which can be as large as you want.

GMNGeoffrey committed:
Still use contigmap even without contigs

I found this helpful for passing through the length and having that get
validated. I can drop it if it seems like a bad idea. I do think it's a
bit weird that we throw away the whole contigmap if contigs isn't set
here. Mabye the idea is that the rest of the parameters don't make sense
without contigs?

GMNGeoffrey committed:
Fix explicit scaffold_list

OmegaConf configs aren't actually dictionaries and their list elements
aren't actually lists (nor do they inherit from them), so this
`isinstance` check wasn't working.

GMNGeoffrey committed:
Expand scaffold guided and partial diffusion configs

Partial diffusion basically has to be done by hand because it's just
different pdb files. Writing it into the Python script seemed like
unnecessary complexity and indirection without any real advantage.

For the scaffold-guided configs I extended the C and N terminus lengths.
I did this even for the `ppi_scaffolded` example so that we could make
it longer as well. Hopefully this still creates a reasonable test case
from a model performance perspective, even if it doesn't make much sense
biologically (which I can't evaluate very well).

GMNGeoffrey committed:
Reorder configs a bit

Also generated comments that I think were a bit duplicative

GMNGeoffrey committed:
Rename configs.yaml -> configs.base.yaml

I'm going to generate configs

GMNGeoffrey committed:
Remove `config_name` from test configs

This causes problems when trying to instead use these for benchmarking
because you can't really use two different configs inheriting from base
(symmetry and benchmark, in this case).

The symmetry config doesn't really specify much and everything is either
overridden by the tests or already matches the base default. Confirmed
this makes no difference to the outputs.

GMNGeoffrey commented:
Ok, I think this is good to go @Joachim-Sh. I extended all the scaffold guided cases using the same strategy and added explicit configs with larger PDBs for partial diffusion

GMNGeoffrey committed:
Merge branch 'brium' into rfpeptide

GMNGeoffrey committed:
Add test cases for new rfpeptide

I'm not really sure that making cyclic binders that long really makes
sense, but it works...

Joachim-Sh commented on rfdiffusion/contigs.py:
```diff
@@ -216,6 +221,8 @@ def get_sampled_mask(self):
                                 subcon_out.append(f"{length_inpaint}-{length_inpaint}")
                                 sampled_mask_length += int(subcon)
                     sampled_mask.append("/".join(subcon_out))
+            max_sampled_length = max(max_sampled_length, sampled_mask_length)
```
What are you checking here?  max_sampled_length has to be larger than 0, and min_sampled_length should be smaller than sys.maxsize?

GMNGeoffrey replied:
It's keeping track of the maximum and minimum sampled length so that we can provide a more helpful error message below. 0 and sys.maxsize are just initialization values. In the first iteration they should be replaced with whatever the sampled length was and then going forward if the sampled length is larger than the max so far, we update the max and if it's smaller than the min so far we update the min, etc. So in the end we know what range of lengths we got and can report it if it doesn't match with the length constraint.

Joachim-Sh reviewed (approved):
Nice work, code looks good! I just had one question.
